app/scripts/background.js

```diff
@@ -176,6 +178,103 @@ const sendReadyMessageToTabs = async () => {
   }
 };
 
+/**
+ * Detects known phishing pages as soon as the browser begins to load the
+ * page. If the page is a known phishing page, the user is redirected to the
+ * phishing warning page.
+ *
+ * This detection works even if the phishing page is now a redirect to a new
+ * domain that our phishing detection system is not aware of.
+ *
+ * @param {MetamaskController} theController
+ */
+function maybeDetectPhishing(theController) {
+  async function redirectTab(tabId, url) {
+    try {
+      return await browser.tabs.update(tabId, {
+        url,
+      });
+    } catch (error) {
+      return sentry?.captureException(error);
+    }
+  }
+  // we can use the blocking API in MV2, but not in MV3
+  const blocking = !isManifestV3;
+  browser.webRequest.onBeforeRequest.addListener(
+    (details) => {
+      if (details.tabId === browser.tabs.TAB_ID_NONE) {
+        return {};
+      }
+
+      const onboardState = theController.onboardingController.store.getState();
+      if (!onboardState.completedOnboarding) {
+        return {};
+      }
+
+      const prefState = theController.preferencesController.store.getState();
+      if (!prefState.usePhishDetect) {
+        return {};
+      }
+
+      // ignore requests that come from our phishing warning page
+      if (
+        details.initiator &&
+        // compare normalized URLs
+        new URL(details.initiator).toString() === phishingPageHref
+      ) {
+        return {};
+      }
+
+      const { hostname, href, searchParams } = new URL(details.url);
+      if (inTest) {
+        if (searchParams.has('IN_TEST_BYPASS_EARLY_PHISHING_DETECTION')) {
+          // this is a test page that needs to bypass early phishing detection
+          return {};
+        }
+      }
```

Comment:
I'm not proud of the `IN_TEST` logic sprinkled in here, but I'm not sure if there is a better way. Let me know if you think of one!

Reply:
It can get into this state naturally. The `browser.webRequest.onBeforeRequest` hook takes a bit before the browser "installs" it on start up. But our `content_script` looks to be injected immediately. This flag is meant to test that case. Under "natural" use this is just a race condition, but we can't rely on that for our tests, so this code block is mean to simulate it, and forces the `content_script` to do the phishing detection.